### PR TITLE
Smart contract security best practices

### PR DESCRIPTION
**Pull request by krofax**
Added a section on smart contract security with some notes on auditing and useful links on smart contract security.

### CODE REVIEW

File: docs/dev/building-on-zksync/contracts/contract-deployment.md [hunk 1]
```diff
@@ -52,6 +52,22 @@ The 32-byte hash of the bytecode of a zkSync contract is calculated in the follo
 - The second 2 bytes denote the length of the bytecode in 32-byte words.
 - The rest of the 28-byte (i.e. 28 low big-endian bytes) are equal to the last 28 bytes of the `sha256` hash of the contract's bytecode.
 
+## Smart contract security
+
+Smart contract security is critical, as a single vulnerability in a smart contract can lead to loss of funds.
```
Comment (kailash-manasarovar) [inline docs/dev/building-on-zksync/contracts/contract-deployment.md:57]:
```suggestion
Smart contract security is critical. A single vulnerability in a smart contract can lead to loss of funds. Make sure your contracts are secure against common threats.
```

File: docs/dev/building-on-zksync/contracts/contract-deployment.md [hunk 2]
```diff
@@ -52,6 +52,22 @@ The 32-byte hash of the bytecode of a zkSync contract is calculated in the follo
 - The second 2 bytes denote the length of the bytecode in 32-byte words.
 - The rest of the 28-byte (i.e. 28 low big-endian bytes) are equal to the last 28 bytes of the `sha256` hash of the contract's bytecode.
 
+## Smart contract security
+
+Smart contract security is critical, as a single vulnerability in a smart contract can lead to loss of funds. 
+
+One common vulnerability in smart contracts is the reentrancy attack, which occurs when a contract's code allows an attacker to repeatedly call the same function before the previous function call has completed.
```
Comment (kailash-manasarovar) [inline docs/dev/building-on-zksync/contracts/contract-deployment.md:59]:
```suggestion
A common Solidity smart contract attack is reentrancy. This threat exploits vulnerabilities in contract code that allow an attacker to repeatedly call the same function that withdraws funds.
```

File: docs/dev/building-on-zksync/contracts/contract-deployment.md [hunk 3]
```diff
@@ -52,6 +52,22 @@ The 32-byte hash of the bytecode of a zkSync contract is calculated in the follo
...
+
+Auditing smart contracts for security holes is crucial to prevent theft or other malicious activities. An audit involves a thorough review of the contract's code and its underlying logic to identify any vulnerabilities or weaknesses that could be exploited by attackers. Auditors look for things like buffer overflows, integer overflows, and other types of security issues that can lead to the loss of assets or other unwanted outcomes. This review process should include both manual and automated testing to ensure that all vulnerabilities are identified.
```
Comment (kailash-manasarovar) [inline docs/dev/building-on-zksync/contracts/contract-deployment.md:61]:
```suggestion
Auditing smart contracts for security holes prevents theft and other malicious activities. An audit involves a thorough review of the contract's code and its underlying logic to identify any vulnerabilities or weaknesses that could be exploited by attackers. Auditors look for things like buffer overflows, integer overflows, and other types of security issues that can lead to the loss of assets or other unwanted outcomes. This review process should include both manual and automated testing to ensure that all vulnerabilities are identified.
```

File: docs/dev/building-on-zksync/contracts/contract-deployment.md [hunk 4]
```diff
@@ -52,6 +52,22 @@ The 32-byte hash of the bytecode of a zkSync contract is calculated in the follo
...
+
+The process of auditing a smart contract should be carried out by experts who have the necessary knowledge and experience to identify potential security risks. Investing in a thorough audit can help prevent security breaches and protect investors and users from losses, reputation damage, and legal issues. Therefore, it's essential to prioritize smart contract security and take proactive measures to ensure that they are thoroughly audited for security holes before deploying your smart contract on zkSync Era network.
+
+The following are some resources on Smart contract security best practices to consider:
```
Comment (kailash-manasarovar) [inline docs/dev/building-on-zksync/contracts/contract-deployment.md:65]:
```suggestion
For detailed information on smart contract vulnerabilities and security best practices, refer to the following resources:
```

File: docs/dev/building-on-zksync/contracts/contract-deployment.md [hunk 5]
```diff
@@ -52,6 +52,22 @@ The 32-byte hash of the bytecode of a zkSync contract is calculated in the follo
...
+
+- (smart contract best practices)[https://consensys.github.io/smart-contract-best-practices/]
```
Comment (kailash-manasarovar) [inline docs/dev/building-on-zksync/contracts/contract-deployment.md:67]:
```suggestion
- (Smart contract best practices)[https://consensys.github.io/smart-contract-best-practices/].
```

File: docs/dev/building-on-zksync/contracts/contract-deployment.md [hunk 6]
```diff
@@ -52,6 +52,22 @@ The 32-byte hash of the bytecode of a zkSync contract is calculated in the follo
...
+
+- (Security considerations)[https://docs.soliditylang.org/en/latest/security-considerations.html]
```
Comment (kailash-manasarovar) [inline docs/dev/building-on-zksync/contracts/contract-deployment.md:69]:
```suggestion
- (Solidity docs security considerations)[https://docs.soliditylang.org/en/latest/security-considerations.html].
```

**Review by kailash-manasarovar — CHANGES_REQUESTED** [review]
Some language changes.

**Comment by krofax** [comment]
> Some language changes.

Updated @kailash-manasarovar

**Review by kailash-manasarovar — APPROVED** [review]

**Pull request merged**